src/test/test.ts

```diff
@@ -145,14 +145,39 @@ export class Suite extends Base implements reporterTypes.Suite {
     return result;
   }
 
-  _hasOnly(): boolean {
+  _getOnlyItems(): (Spec | Suite)[] {
+    const items: (Spec | Suite)[] = [];
     if (this._only)
-      return true;
-    if (this.suites.find(suite => suite._hasOnly()))
-      return true;
-    if (this.specs.find(spec => spec._only))
-      return true;
-    return false;
+      items.push(this);
+    for (const suite of this.suites)
+      items.push(...suite._getOnlyItems());
+    items.push(...this.specs.filter(spec => spec._only));
+    return items;
+  }
+
+  _getUniqueItemsPerFile(): (Spec | Suite)[] {
```

Comment:
I'd move this function to `runner.ts`, because it only works for top-level suite.